ChopChop/Storage/StorageManager.swift

```diff
@@ -228,6 +233,162 @@ extension StorageManager {
+        }
+        let steps = recipe.steps.map({ $0.content })
+        let ingredients = recipe.ingredients.map({
+            OnlineIngredientRecord(name: $0.name, quantity: $0.quantity.record)
+        })
+        let recipeRecord = OnlineRecipeRecord(
+            id: recipe.onlineId,
+            name: recipe.name,
+            creator: userId,
+            servings: recipe.servings,
+            cuisine: cuisine,
+            difficulty: recipe.difficulty,
+            ingredients: ingredients,
+            steps: steps
+        )
+        try firebase.updateRecipeDetails(recipe: recipeRecord)
+    }
+
+    func removeRecipeFromOnline(recipe: OnlineRecipe) throws {
+
+        try firebase.removeRecipe(recipeId: recipe.id)
+        for rating in recipe.ratings {
+            firebase.removeUserRecipeRating(userId: rating.userId, rating: UserRating(recipeOnlineId: recipe.id, score: rating.score))
+        }
+    }
+
+//    func removeRecipeFromLocal(recipeId: String) throws {
+//
+//    }
+
+    func addFollowee(userId: String, followeeId: String) {
+        firebase.addFollowee(userId: userId, followeeId: followeeId)
+    }
+
+    func removeFollowee(userId: String, followeeId: String) {
+        firebase.removeFollowee(userId: userId, followeeId: followeeId)
+    }
+
+    func rateRecipe(recipeId: String, userId: String, rating: RatingScore) throws {
+        firebase.addUserRecipeRating(userId: userId, rating: UserRating(recipeOnlineId: recipeId, score: rating))
+        firebase.addRecipeRating(onlineRecipeId: recipeId, rating: RecipeRating(userId: userId, score: rating))
+//        let ratingRecord = RatingRecord(userId: recipeId, recipeId: userId, rating: rating)
+//        try firebase.addRating(rating: ratingRecord)
+    }
+
+//    func unrateRecipe(ratingId: String) {
+//        firebase.removeRating(ratingId: ratingId)
+//    }
+    func unrateRecipe(recipeId: String, rating: RecipeRating) {
+        firebase.removeRecipeRating(onlineRecipeId: recipeId, rating: rating)
+        firebase.removeUserRecipeRating(userId: rating.userId, rating: UserRating(recipeOnlineId: recipeId, score: rating.score))
+    }
+
+//    func rerateRecipe(ratingId: String, newRating: RatingScore) {
+//        firebase.updateRating(ratingId: ratingId, rating: newRating)
+//    }
+    func rerateRecipe(recipeId: String, newRating: RecipeRating) {
+        firebase.updateRecipeRating(userId: newRating.userId, recipeId: recipeId, newScore: newRating.score)
+        firebase.updateUserRating(userId: newRating.userId, recipeId: recipeId, newScore: newRating.score)
+    }
+
+    func fetchAllFriends(userId: String) -> AnyPublisher<[User], Error> {
+        firebase.fetchFriendsId(userId: userId)
+            .flatMap({ followees in
+                firebase.fetchUsers(userId: followees)
+            })
+            .eraseToAnyPublisher()
+    }
+
+    func fetchAllSelfPublishedRecipes(userId: String) -> AnyPublisher<[OnlineRecipe], Error> {
```

Comment:
dont need duplicate, just generalize a function to take in an array of userIds

Reply:
IF got time: can sort by create date?

Add a new field to OnlineRecipe to store create date
https://stackoverflow.com/questions/51846914/add-timestamp-in-firestore-documents
https://stackoverflow.com/questions/47967868/firebase-store-server-timestamp-when-document-created/47969509